### PR TITLE
Make data shuffling in `run_clm_flax.py` respect global seed

### PR DESCRIPTION
**Pull request by bminixhofer**
# What does this PR do?

Use `jax.random.permutation` instead of `np.random.permutation` in the `data_loader` function of `run_clm_flax.py` to make it use the global seed. Currently batch order would probably vary across runs, regardless of the global seed.

Also changes `np.arange` to `jnp.arange` and `np.array` to `jnp.array` and removes the numpy import, although that would not be strictly necessary.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patil-suraj

### CODE REVIEW

**Comment by patil-suraj** [comment]
Thanks a lot for the PR @bminixhofer, good catch!

However, initially, we actually used `jax.random.permutation` and `jax.arange` in the data loader but then switched to numpy as we observed it was causing some issues with JAX's asynchronous dispatch. Since JAX by default puts everything on the device it could cause some issues (especially on TPU) if the dataloader /collator is used with multiple threads to do background fetching. This also leads to major slowdowns. So all flax examples now don't use JAX functions in pre-processing, loading/collating etc. With this, the TPU can be busy all the time doing the actual computation and won't be blocked by processing and loading.

But I see the problem, so maybe we could use the seed with numpy to make data shuffling reproducible.

cc @patrickvonplaten .

**Comment by bminixhofer** [comment]
Interesting, what about `run_mlm_flax.py`? I was having a look prior to this PR, and it seems `jax.random.permutation` is also used for data shuffling, or am I missing something?

https://github.com/huggingface/transformers/blob/76c4d8bf26de3e4ab23b8afeed68479c2bbd9cbd/examples/flax/language-modeling/run_mlm_flax.py#L624-L626

**Comment by patil-suraj** [comment]
That should also be changed. All flax examples are still not completely consistent with each other, which needs to be fixed.

File: examples/flax/language-modeling/run_clm_flax.py [hunk 1]
```diff
@@ -187,16 +186,16 @@ def data_loader(rng: jax.random.PRNGKey, dataset: Dataset, batch_size: int, shuf
...
-        batch_idx = np.arange(len(dataset))
+        batch_idx = jnp.arange(len(dataset))
 
     batch_idx = batch_idx[: steps_per_epoch * batch_size]  # Skip incomplete batch.
     batch_idx = batch_idx.reshape((steps_per_epoch, batch_size))
 
     for idx in batch_idx:
         batch = dataset[idx]
-        batch = {k: np.array(v) for k, v in batch.items()}
+        batch = {k: jnp.array(v) for k, v in batch.items()}
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:198]: this should stay `np`

File: examples/flax/language-modeling/run_clm_flax.py [hunk 2]
```diff
@@ -187,16 +186,16 @@ def data_loader(rng: jax.random.PRNGKey, dataset: Dataset, batch_size: int, shuf
     steps_per_epoch = len(dataset) // batch_size
 
     if shuffle:
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_clm_flax.py:188]: @patil-suraj - I think here we could actually switch to `jax.random.permutation` as this is called only once per epoch. I'm pretty sure that it wouldn't slow down the training

File: examples/flax/language-modeling/run_clm_flax.py (hunk 1 above)
Comment (patrickvonplaten) [reply examples/flax/language-modeling/run_clm_flax.py:198]: due to asynchronous dispatch as noted by @patil-suraj

Comment (bminixhofer) [reply examples/flax/language-modeling/run_clm_flax.py]: Ok, I switched it back :+1:

**Comment by bminixhofer** [comment]
@patil-suraj @patrickvonplaten Is this something you want to change or should I close this PR?

**Comment by patrickvonplaten** [comment]
Hey @bminixhofer - sorry for being so slow here. @patil-suraj I'm happy to merge the PR. Think it's good to have 100% reproduciblity with JAX's random seed. I don't think this slows down the script as it's called just once per epoch. If you're ok with the changes feel free to merge

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
ping @patil-suraj again

**Comment by patrickvonplaten** [comment]
ping @patil-suraj again

**Review by patil-suraj — APPROVED** [review]
Sorry about being super slow here. Agree with @patrickvonplaten and think we can change this to use `jax.random`.
also, note that even with numpy batch order will be preserved since we set the seed for numpy as well.
cf https://github.com/huggingface/transformers/blob/546a91abe95117fb06d3ed34edfa0b8010d9b48c/examples/flax/language-modeling/run_clm_flax.py#L342

@bminixhofer Thank you for fixing this!

**Comment by bminixhofer** [comment]
Thanks for merging this! I vaguely remember having problems with batch order with the code as it was previously, but I am not completely sure (it's been some time :sweat_smile: ).
